lib/common/optional/types.go

```diff
@@ -146,6 +146,80 @@ func (x *Bytes) Decode(r io.Reader) (*Bytes, error) {
 	return x, nil
 }
 
+// Boolean represents an optional bool type.
+type Boolean struct {
+	exists bool
+	value  bool
+}
+
+// NewBoolean returns a new optional.Boolean
+func NewBoolean(exists bool, value bool) *Boolean {
+	return &Boolean{
+		exists: exists,
+		value:  value,
+	}
+}
+
+// Exists returns true if the value is Some, false if it is None.
+func (x *Boolean) Exists() bool {
+	return x.exists
+}
+
+// Value returns the []byte value. It returns nil if it is None.
+func (x *Boolean) Value() bool {
+	return x.value
+}
+
+// Set sets the exists and value fields.
+func (x *Boolean) Set(value bool) {
+	x.exists = true
+	x.value = value
+}
+
+// Encode returns the SCALE encoded optional
+func (x *Boolean) Encode() ([]byte, error) {
+	if !x.exists {
+		return []byte{0}, nil
+	}
+	var encodeValue []byte
+
+	if !x.exists {
+		encodeValue = []byte{0}
+	} else if !x.value {
+		encodeValue = []byte{1}
+	} else {
+		encodeValue = []byte{2}
+	}
+
+	return encodeValue, nil
+}
+
+// Decode return an optional Boolean from scale encoded data
+func (x *Boolean) Decode(r io.Reader) (*Boolean, error) {
+	decoded, err := common.ReadByte(r)
+
+	if err != nil {
+		return nil, errors.New("Decoding failed, invalid optional")
+	}
+
+	if decoded > 2 {
+		return nil, errors.New("Decoding failed, invalid optional")
+	}
+
+	if decoded == 0 {
+		x.exists = false
+		x.value = false
+	} else if decoded == 1 {
+		x.exists = true
+		x.value = false
+	} else {
+		x.exists = true
+		x.value = true
+	}
```

Comment:
I think these 2 are switched, 1 is true and 2 is false

also, you should explicity check that it's 0, 1, or 2, otherwise return an error saying can't decode invalid value or something along those lines

Reply:
@noot There's a check above that rejects values larger than 2